okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -170,32 +172,44 @@ private RealConnection findConnection(int connectTimeout, int readTimeout, int w
         refusedStreamCount = 0;
       }
     }
-    RealConnection newConnection = new RealConnection(connectionPool, selectedRoute);
+
+    RealConnection result = new RealConnection(connectionPool, selectedRoute);
 
     synchronized (connectionPool) {
-      acquire(newConnection);
-      Internal.instance.put(connectionPool, newConnection);
-      this.connection = newConnection;
+      acquire(result);
+      Internal.instance.put(connectionPool, result);
```

Comment:
Yeah good thinking. Here's the moving parts currently being balanced:

 * Keeping `streamAllocation.connection` in sync with `connection.allocations`.
 * Assigning `streamAllocation.connection` before the connect completes so that it can be canceled in flight. 
 * Not leaking anything, and detecting when we do.

I’d like to try and force the bad interleavings, maybe with a torture test. I'll see how that goes.